### PR TITLE
Creating default networkpolicies on CNF namespace creation

### PR DESCRIPTION
**Pull request by nsilla**
The CNF certification test suite fails in test **networking-network-policy-deny-all** with recommended mitigation:

_"Ensure that a NetworkPolicy with a default deny-all is applied. After the default is applied, apply a network policy to allow the traffic your application requires."_

An operator may not ship network policy manifest, so it's left for the deploying agent/code to create the mandatory network policies.

### CODE REVIEW

**Comment by nsilla** [comment]
To test:

```
dci-pipeline-check https://github.com/dci-labs/example-cnf-config/pull/93 ocp-4.14-vanilla example-cnf \
example-cnf:ansible_extravars=test_network_function_version:HEAD \
example-cnf:ansible_extravars=numa_aware_topology:topo-aware-scheduler
```

**Comment by ramperher** [comment]
@nsilla , some questions I have:

- Do we intend to create one ingress-egress network policy per operator (similarly to what we currently use on tnf_test_example)? Or the same for all operators? Because the current ones are named "testpmd-operator" and that's a bit confusing for me, it's like they were just used for testpmd-operator, but in its definition, they seem to be related to all example-cnf namespace resources
- Why do you need to use tnf HEAD version?

**Comment by ramperher** [comment]
@nsilla , two aspects I am observing here:

- the testpmd-operator-controller-manager complains about this in its logs, maybe something related to traffic denied because of the network policies that you're implementing?

```
{"level":"error","ts":"2024-03-22T19:31:36Z","logger":"controller-runtime.object-cache.setup","msg":"Failed to get API Group-Resources","error":"Get \"https://172.30.0.1:443/api\": dial tcp 172.30.0.1:443: i/o timeout","stacktrace":"sigs.k8s.io/controller-runtime/pkg/cache.defaultOpts\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/cache/cache.go:266\nsigs.k8s.io/controller-runtime/pkg/cache.New\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/cache/cache.go:206\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.run.func1\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:183\nsigs.k8s.io/controller-runtime/pkg/cluster.New\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/cluster/cluster.go:225\nsigs.k8s.io/controller-runtime/pkg/manager.New\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/manager/manager.go:401\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.run\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:188\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.NewCmd.func1\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:81\ngithub.com/spf13/cobra.(*Command).execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\tansible-operator-plugins/cmd/ansible-operator/main.go:40\nruntime.main\n\t/opt/hostedtoolcache/go/1.20.11/x64/src/runtime/proc.go:250"}
{"level":"error","ts":"2024-03-22T19:31:36Z","logger":"cmd","msg":"Failed to create a new manager.","Namespace":"","error":"could not create RESTMapper from config: Get \"https://172.30.0.1:443/api\": dial tcp 172.30.0.1:443: i/o timeout","stacktrace":"github.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.run\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:190\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.NewCmd.func1\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:81\ngithub.com/spf13/cobra.(*Command).execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\tansible-operator-plugins/cmd/ansible-operator/main.go:40\nruntime.main\n\t/opt/hostedtoolcache/go/1.20.11/x64/src/runtime/proc.go:250"}
```

- Linked to that, I cannot see any podSelector clause specified in your network policies, I fear you're denying all traffic by default and not matching the "allow" rules with the pods from example-cnf

**Comment by manurodriguez** [comment]
I see the policies but I'm blocking something that the CSV is not able to get successfully installed:
```
[kni@provisioner.cluster6.dfwt5g.lab ~]$ oc -n example-cnf get NetworkPolicies
NAME                                                 POD-SELECTOR                   AGE
allow-dns-egress                                     <none>                         22s
allow-dns-ingress                                    <none>                         22s
allow-intra-namespace-example-cnf-type-cnf-app       example-cnf-type=cnf-app       25s
allow-intra-namespace-example-cnf-type-lb-app        example-cnf-type=lb-app        26s
allow-intra-namespace-example-cnf-type-pkt-gen       example-cnf-type=pkt-gen       24s
allow-intra-namespace-example-cnf-type-pkt-gen-app   example-cnf-type=pkt-gen-app   23s
allow-istio-egress                                   <none>                         22s
allow-istio-ingress                                  <none>                         22s
default-deny-egress-example-cnf-type-cnf-app         example-cnf-type=cnf-app       25s
default-deny-egress-example-cnf-type-lb-app          example-cnf-type=lb-app        26s
default-deny-egress-example-cnf-type-pkt-gen         example-cnf-type=pkt-gen       24s
default-deny-egress-example-cnf-type-pkt-gen-app     example-cnf-type=pkt-gen-app   23s
default-deny-ingress-example-cnf-type-cnf-app        example-cnf-type=cnf-app       25s
default-deny-ingress-example-cnf-type-lb-app         example-cnf-type=lb-app        26s
default-deny-ingress-example-cnf-type-pkt-gen        example-cnf-type=pkt-gen       24s
default-deny-ingress-example-cnf-type-pkt-gen-app    example-cnf-type=pkt-gen-app   23s
from-openshift-dns                                   <none>                         22s
to-openshift-dns                                     <none>                         22s

[kni@provisioner.cluster3.dfwt5g.lab ~]$ oc -n example-cnf get pods
NAME                                                       READY   STATUS      RESTARTS   AGE
cnf-app-mac-operator-controller-manager-7d766f69f8-5sn5z   1/1     Running     0          28m
cnf-app-mac-operator-controller-manager-7d766f69f8-6dlqv   1/1     Running     0          28m
job-trex-app-mzt8f                                         0/1     Completed   0          22m
testpmd-app-68767d96c7-tmd2s                               1/1     Running     0          24m
testpmd-operator-controller-manager-b758584f8-7lcpp        1/1     Running     0          26m
testpmd-operator-controller-manager-b758584f8-zgv5p        1/1     Running     0          26m
trex-operator-controller-manager-964659fb4-hbtdm           1/1     Running     0          25m
trex-operator-controller-manager-964659fb4-l2tn6           1/1     Running     0          25m
trexconfig-c9bc5b9f6-2f88h                                 1/1     Running     0          23m

[kni@provisioner.cluster6.dfwt5g.lab ~]$ oc -n example-cnf get csv                                                                                                                                                                           
NAME                                                DISPLAY                      VERSION                       REPLACES   PHASE                                                                                                              
cnf-app-mac-operator.v0.2.16-202405211327.4010727   CNF App MAC Fetch Operator   0.2.16-202405211327.4010727              Succeeded                                                                                                          
testpmd-operator.v0.2.15-202405211327.4010727       TestPMD Operator             0.2.15-202405211327.4010727              Installing                                                                                                         
                                                                                                             
[kni@provisioner.cluster6.dfwt5g.lab ~]$ oc -n example-cnf get csv testpmd-operator.v0.2.15-202405211327.4010727 -o json | jq .status.conditions  
...
  {
    "lastTransitionTime": "2024-06-25T04:46:19Z",
    "lastUpdateTime": "2024-06-25T04:46:19Z",
    "message": "installing: waiting for deployment testpmd-operator-controller-manager to become ready: waiting for spec update of deployment \"testpmd-operator-controller-manager\" to be observed...",
    "phase": "Installing",
    "reason": "InstallWaiting"
  },
  {
    "lastTransitionTime": "2024-06-25T04:51:18Z",
    "lastUpdateTime": "2024-06-25T04:51:18Z",
    "message": "install timeout",
    "phase": "Failed",
    "reason": "InstallCheckFailed"
  },
  {
    "lastTransitionTime": "2024-06-25T04:51:19Z",
    "lastUpdateTime": "2024-06-25T04:51:19Z",
    "message": "installing: waiting for deployment testpmd-operator-controller-manager to become ready: deployment \"testpmd-operator-controller-manager\" not available: Deployment does not have minimum availability.",                   
    "phase": "Pending",
    "reason": "NeedsReinstall"
  },
...
```
I'll keep digging tomorrow,

File: testpmd/hooks/templates/network_policies_am_traffic.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                        kubernetes.io/metadata.name: istio-system
+            egress:
+              - to:
+                  - namespaceSelector:
+                      matchLabels:
+                        kubernetes.io/metadata.name: "{{ cnf_namespace }}"
+        - kind: NetworkPolicy
+          apiVersion: networking.k8s.io/v1
+          metadata:
+            name: allow-istio-egress
```
[the remark this answers is not in the document]
Comment (manurodriguez) [reply testpmd/hooks/templates/network_policies_am_traffic.yml.j2]: ups, my bad, wrong copy/paste, I removed it

File: testpmd/hooks/templates/network_policies_am_traffic.yml.j2
[the remark this answers is not in the document]
Comment (manurodriguez) [reply testpmd/hooks/templates/network_policies_am_traffic.yml.j2:1]: done, thanks

File: testpmd/hooks/templates/network_policies_am_traffic.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,112 @@
+      apiVersion: v1
+      kind: List
+      items:
+        - apiVersion: networking.k8s.io/v1
+          kind: NetworkPolicy
+          metadata:
+            name: to-openshift-dns
```
[the remark this answers is not in the document]
Comment (manurodriguez) [reply testpmd/hooks/templates/network_policies_am_traffic.yml.j2]: that's true, I removed it

File: testpmd/hooks/templates/network_policies_am_traffic.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            podSelector: {}
+            egress:
+              - to:
+                  - namespaceSelector:
+                      matchLabels:
+                        kubernetes.io/metadata.name: openshift-dns
+        - apiVersion: networking.k8s.io/v1
+          kind: NetworkPolicy
+          metadata:
+            name: from-openshift-dns
```
[the remark this answers is not in the document]
Comment (manurodriguez) [reply testpmd/hooks/templates/network_policies_am_traffic.yml.j2]: removed the duplicate

File: testpmd/hooks/pre-run.yml [hunk 4]
```diff
@@ -119,6 +119,23 @@
   set_fact:
     ns_list: "{{ ns_list|default([]) + [ cnf_namespace ] }}"
 
+- name: Set the CNF mandatory network policies
+  community.kubernetes.k8s:
+    state: present
+    definition: "{{ lookup('template', 'templates/network_policies_cnf_traffic.yml.j2') | from_yaml }}"
+  loop:
```
[the remark this answers is not in the document]
Comment (manurodriguez) [reply testpmd/hooks/pre-run.yml:126]: agreed,  I was checking the `targetpodlabels` in tnf_config of example-cnf-pipeline.yml and I saw that label too, but I wasn't sure to add it. Thanks

**Comment by manurodriguez** [comment]
> I've been investigating a little bit on this, and I think any of these two namespaces should be taking care of the requests to k8s API: `openshift-kube-apiserver` (I think it's this one because there's one pod per master node, and as you know, the k8s API server functionality is provided by master nodes) and `openshift-apiserver`. So maybe I'll propose to add some network policies to allow traffic from example-cnf namespace to these namespaces.

Thanks for digging Ramon, let me give it a try, I'll add first one of those NS to see if we can catch exactly which one is needed, maybe is both : )

**Comment by manurodriguez** [comment]
First try with only allowing openshift-apiserver didn't work, same error:
```
[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get pods
NAME                                                       READY   STATUS    RESTARTS      AGE
cnf-app-mac-operator-controller-manager-7d66d79d9b-5nxsq   1/1     Running   0             15m
cnf-app-mac-operator-controller-manager-7d66d79d9b-96bdr   1/1     Running   0             14m
testpmd-lb-operator-controller-manager-66cbd6b8cf-9g5kk    0/1     Running   4 (10s ago)   2m7s
testpmd-lb-operator-controller-manager-66cbd6b8cf-kptsk    0/1     Running   3 (16s ago)   106s

[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get csv                             
NAME                                                DISPLAY                      VERSION                       REPLACES               PHASE
cnf-app-mac-operator.v0.2.16-202405211327.4010727   CNF App MAC Fetch Operator   0.2.16-202405211327.4010727                          Succeeded
testpmd-lb-operator.v0.2.14-202405211327.4010727    LoadBalancer                 0.2.14-202405211327.4010727                          Installing

[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf describe pod testpmd-lb-operator-controller-manager-66cbd6b8cf-9g5kk
...
Events:
  Type     Reason          Age                From               Message
  ----     ------          ----               ----               -------
  Normal   Scheduled       53s                default-scheduler  Successfully assigned example-cnf/testpmd-lb-operator-controller-manager-66cbd6b8cf-9g5kk to worker-0                                                                       
  Normal   AddedInterface  52s                multus             Add eth0 [10.131.0.81/23 fd02:0:0:4::51/64] from ovn-kubernetes                                                                                                             
  Normal   Pulled          24s (x2 over 52s)  kubelet            Container image "quay.io/rh-nfv-int/testpmd-lb-operator@sha256:1b212325eb26c0302638b602e7aee3e9bc9806fb862512d07f328fa4758cb921" already present on machine                 
  Normal   Created         24s (x2 over 52s)  kubelet            Created container manager
  Normal   Started         24s (x2 over 52s)  kubelet            Started container manager
  Normal   Killing         24s                kubelet            Container manager failed startup probe, will be restarted                                                                                                                   
  Warning  Unhealthy       4s (x5 over 44s)   kubelet            Startup probe failed: Get "http://10.131.0.81:6789/healthz": dial tcp 10.131.0.81:6789: connect: connection refused 
```

In the log testpmd-lb-operator-controller-manager-66cbd6b8cf-2dsbv.log  I see the same error:
```
{"level":"error","ts":"2024-06-25T19:20:42Z","logger":"controller-runtime.object-cache.setup","msg":"Failed to get API Group-Resources","error":"Get \"https://172.30.0.1:443/api\": dial tcp 172.30.0.1:443: i/o timeout","stacktrace":"sigs.k8s.io/controller-runtime/pkg/cache.defaultOpts\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/cache/cache.go:266\nsigs.k8s.io/controller-runtime/pkg/cache.New\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/cache/cache.go:206\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.run.func1\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:183\nsigs.k8s.io/controller-runtime/pkg/cluster.New\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/cluster/cluster.go:225\nsigs.k8s.io/controller-runtime/pkg/manager.New\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.3/pkg/manager/manager.go:401\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.run\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:188\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.NewCmd.func1\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:81\ngithub.com/spf13/cobra.(*Command).execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\tansible-operator-plugins/cmd/ansible-operator/main.go:40\nruntime.main\n\t/opt/hostedtoolcache/go/1.20.11/x64/src/runtime/proc.go:250"}
{"level":"error","ts":"2024-06-25T19:20:42Z","logger":"cmd","msg":"Failed to create a new manager.","Namespace":"","error":"could not create RESTMapper from config: Get \"https://172.30.0.1:443/api\": dial tcp 172.30.0.1:443: i/o timeout","stacktrace":"github.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.run\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:190\ngithub.com/operator-framework/ansible-operator-plugins/internal/cmd/ansible-operator/run.NewCmd.func1\n\tansible-operator-plugins/internal/cmd/ansible-operator/run/cmd.go:81\ngithub.com/spf13/cobra.(*Command).execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\t/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\tansible-operator-plugins/cmd/ansible-operator/main.go:40\nruntime.main\n\t/opt/hostedtoolcache/go/1.20.11/x64/src/runtime/proc.go:250"}
```
And actually the log cnf-app-mac-operator-controller-manager-7d66d79d9b-5nxsq.log is also showing some errors:
```
E0625 19:11:39.426955       1 leaderelection.go:327] error retrieving resource lock example-cnf/69868ffe.openshift.io: Get "https://172.30.0.1:443/apis/coordination.k8s.io/v1/namespaces/example-cnf/leases/69868ffe.openshift.io": dial tcp 172.30.0.1:443: i/o timeout
```
I'll add a Net Policy fir the other api NS and keep testing

**Comment by manurodriguez** [comment]
same error even with a policy for openshift-kube-apiserver, but I found a very nice feature to log Network Policy events https://docs.openshift.com/container-platform/4.15/networking/ovn_kubernetes_network_provider/logging-network-policy.html#logging-network-policy

Then I was able to see that we were still blocking the egress for the API TCP 6443
```
[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc annotate namespace example-cnf k8s.ovn.org/acl-logging='{ "deny": "alert", "allow": "notice" }'                                                                                                  
namespace/example-cnf annotated

[kni@provisioner.cluster5.dfwt5g.lab ~]$ for pod in $(oc get pods -n openshift-ovn-kubernetes -l app=ovnkube-node --no-headers=true | awk '{ print $1 }') ; do     oc exec -it $pod -n openshift-ovn-kubernetes -- tail -4 /var/log/ovn/acl-au
dit-log.log;   done
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
2024-06-25T21:29:34.009Z|00032|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:Egress", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.21,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=43692,tp_dst=6443,tcp_flags=syn
2024-06-25T21:29:36.056Z|00033|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:Egress", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.21,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=43692,tp_dst=6443,tcp_flags=syn
2024-06-25T21:29:40.088Z|00034|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:Egress", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.21,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=43692,tp_dst=6443,tcp_flags=syn
2024-06-25T21:29:48.408Z|00035|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:Egress", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.21,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=43692,tp_dst=6443,tcp_flags=synx`
```
I deleted the old egress policies for the API and created a new policy to allow that traffic
```
[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf delete NetworkPolicy allow-apiserver-egress                                                                                                                                       
networkpolicy.networking.k8s.io "allow-apiserver-egress" deleted

[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf delete NetworkPolicy allow-kube-apiserver-egress
networkpolicy.networking.k8s.io "allow-kube-apiserver-egress" deleted

[kni@provisioner.cluster5.dfwt5g.lab ~]$ vi manny/net-policy-6443-tcp.yml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-api-egress
  namespace: example-cnf
spec:
  podSelector: {}
  egress:
    - ports:
        - protocol: TCP
          port: 6443


[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc apply -f manny/net-policy-6443-tcp.yml                                                                                                                                                           
networkpolicy.networking.k8s.io/allow-api-egress created
```
Now we can see all pod pods running and CSV too, also the logs show the TCP 6443 is permitted:
```
[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get pods
NAME                                                       READY   STATUS    RESTARTS       AGE
cnf-app-mac-operator-controller-manager-7d66d79d9b-7prxv   1/1     Running   0              175m
cnf-app-mac-operator-controller-manager-7d66d79d9b-s2p5l   1/1     Running   0              174m
testpmd-lb-operator-controller-manager-66cbd6b8cf-4jhfb    1/1     Running   0              19m
testpmd-lb-operator-controller-manager-66cbd6b8cf-rt8vj    1/1     Running   48 (25m ago)   168m

[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get csv
NAME                                                DISPLAY                      VERSION                       REPLACES               PHASE
cnf-app-mac-operator.v0.2.16-202405211327.4010727   CNF App MAC Fetch Operator   0.2.16-202405211327.4010727                          Succeeded
testpmd-lb-operator.v0.2.14-202405211327.4010727    LoadBalancer                 0.2.14-202405211327.4010727                          Succeeded

[kni@provisioner.cluster5.dfwt5g.lab ~]$ for pod in $(oc get pods -n openshift-ovn-kubernetes -l app=ovnkube-node --no-headers=true | awk '{ print $1 }') ; do     oc exec -it $pod -n openshift-ovn-kubernetes -- tail -4 /var/log/ovn/acl-au
dit-log.log;   done
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
Defaulted container "ovn-controller" out of: ovn-controller, ovn-acl-logging, kube-rbac-proxy-node, kube-rbac-proxy-ovn-metrics, northd, nbdb, sbdb, ovnkube-controller, kubecfg-setup (init)                                                
2024-06-25T23:27:21.098Z|05390|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:allow-api-egress:Egress:0", verdict=allow, severity=notice, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.22,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46774,tp_dst=6443,tcp_flags=ack
2024-06-25T23:27:21.098Z|05391|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:allow-api-egress:Egress:0", verdict=allow, severity=notice, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.22,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46774,tp_dst=6443,tcp_flags=psh|ack
2024-06-25T23:27:21.098Z|05392|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:allow-api-egress:Egress:0", verdict=allow, severity=notice, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.22,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46774,tp_dst=6443,tcp_flags=psh|ack
2024-06-25T23:27:21.102Z|05393|acl_log(ovn_pinctrl0)|INFO|name="NP:example-cnf:allow-api-egress:Egress:0", verdict=allow, severity=notice, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:81:03:41,dl_dst=0a:58:0a:81:02:01,nw_src=10.129.3.65,nw_dst=192.168.52.22,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46774,tp_dst=6443,tcp_flags=ack
```
I'll update the policies and test again

**Comment by manurodriguez** [comment]
@ramperher we managed to make it work : )
https://www.distributed-ci.io/jobs/328f1aae-4061-4f04-b97b-9b12cf83abed/tests/b390598a-c458-4acb-bfdb-829fc3b96ba7
Screenshot from 2024-06-25 22-37-38

```
[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get NetworkPolicy
NAME                                                     POD-SELECTOR                       AGE
allow-apiserver-egress                                   <none>                             14m
allow-dns-egress                                         <none>                             14m
allow-dns-ingress                                        <none>                             14m
allow-intra-namespace-control-plane-controller-manager   control-plane=controller-manager   14m
allow-intra-namespace-example-cnf-type-cnf-app           example-cnf-type=cnf-app           14m
allow-intra-namespace-example-cnf-type-lb-app            example-cnf-type=lb-app            14m
allow-intra-namespace-example-cnf-type-pkt-gen           example-cnf-type=pkt-gen           14m
allow-intra-namespace-example-cnf-type-pkt-gen-app       example-cnf-type=pkt-gen-app       14m
allow-istio-egress                                       <none>                             14m
allow-istio-ingress                                      <none>                             14m
default-deny-egress-control-plane-controller-manager     control-plane=controller-manager   14m
default-deny-egress-example-cnf-type-cnf-app             example-cnf-type=cnf-app           14m
default-deny-egress-example-cnf-type-lb-app              example-cnf-type=lb-app            14m
default-deny-egress-example-cnf-type-pkt-gen             example-cnf-type=pkt-gen           14m
default-deny-egress-example-cnf-type-pkt-gen-app         example-cnf-type=pkt-gen-app       14m
default-deny-ingress-control-plane-controller-manager    control-plane=controller-manager   14m
default-deny-ingress-example-cnf-type-cnf-app            example-cnf-type=cnf-app           14m
default-deny-ingress-example-cnf-type-lb-app             example-cnf-type=lb-app            14m
default-deny-ingress-example-cnf-type-pkt-gen            example-cnf-type=pkt-gen           14m
default-deny-ingress-example-cnf-type-pkt-gen-app        example-cnf-type=pkt-gen-app       14m

[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get pods
NAME                                                       READY   STATUS    RESTARTS   AGE
cnf-app-mac-operator-controller-manager-7d66d79d9b-bkbr7   1/1     Running   0          9m21s
cnf-app-mac-operator-controller-manager-7d66d79d9b-q5vxm   1/1     Running   0          8m41s
job-trex-app-64tck                                         1/1     Running   0          89s
loadbalancer-6f5b49c9c9-bwzfm                              2/2     Running   0          4m9s
testpmd-app-56d8d87b86-26cjj                               1/1     Running   0          3m16s
testpmd-app-56d8d87b86-zb2zn                               1/1     Running   0          3m16s
testpmd-lb-operator-controller-manager-66cbd6b8cf-2ln7j    1/1     Running   0          7m2s
testpmd-lb-operator-controller-manager-66cbd6b8cf-9c97s    1/1     Running   0          7m13s
testpmd-operator-controller-manager-7459cd99b4-r9gg7       1/1     Running   0          5m57s
testpmd-operator-controller-manager-7459cd99b4-rwzf2       1/1     Running   0          6m8s
trex-operator-controller-manager-7dc6954d58-75kdc          1/1     Running   0          5m1s
trex-operator-controller-manager-7dc6954d58-mt5j8          1/1     Running   0          4m51s
trexconfig-84dd487c5f-xwlmt                                1/1     Running   0          2m23s

[kni@provisioner.cluster5.dfwt5g.lab ~]$ oc -n example-cnf get csv
NAME                                                DISPLAY                      VERSION                       REPLACES               PHASE                                                                                                  
cnf-app-mac-operator.v0.2.16-202405211327.4010727   CNF App MAC Fetch Operator   0.2.16-202405211327.4010727                          Succeeded                                                                                                                                                                                        
testpmd-lb-operator.v0.2.14-202405211327.4010727    LoadBalancer                 0.2.14-202405211327.4010727                          Succeeded                                                                                              
testpmd-operator.v0.2.15-202405211327.4010727       TestPMD Operator             0.2.15-202405211327.4010727                          Succeeded                                                                                              
trex-operator.v0.2.17-202405211327.4010727          TRex Operator                0.2.17-202405211327.4010727                          Succeeded  
```

**Pull request merged**